Include/cpython/objimpl.h

```diff
@@ -97,6 +97,8 @@ typedef struct {
 #define _PyGC_SET_FINALIZED(o) \
     _PyGCHead_SET_FINALIZED(_Py_AS_GC(o))
 
+#define _PyObject_GC_IS_COLLECTING(op) \
+    ((((PyGC_Head*)op)->_gc_prev & _PyGC_PREV_MASK_COLLECTING) != 0)
```

Comment:
Does this macro expect a PyObject* or a PyGC_Head*? In gc_is_collecting() you pass PyGC_Head*, in handle_callback you pass a PyObject*.

Usually, _PyObject_xxx() functions expects a Python object. I suggest to move PyObject_IS_GC() check into _PyObject_GC_IS_COLLECTING() and convert it to a regular function. You can define the function in gcmodule.c.